docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,628 @@
+
+It’s also quite easy to understand what’s happening in a service at a glance, since all the steps are listed sequentially. There’s also a matching system to handle the possible outcomes in a deterministic way, here again you can understand at a glance what a controller will do, for example.
+
+## Getting started
+
+Here’s a simplified service to update a user’s username which demonstrates most available steps:
+```ruby
+class User::UpdateUsername
+  include Service::Base
+
+  params do
+    attribute :id, :integer
+    attribute :username, :string
+
+    validates :id, presence: true
+    validates :username, presence: true, format: { with: /\A[a-zA-Z0-9]+\z/ }
+  end
+  model :user
+  policy :can_update_username
+  transaction do
+    step :update
+    step :log
+  end
+
+  private
+
+  def fetch_user(params:)
+    User.find_by(id: params.id)
+  end
+
+  def can_update_username(guardian:, user:)
+    guardian.can_edit_username?(user)
+  end
+
+  def update(params:, user:)
+    user.update!(username: params.username)
+  end
+
+  def log(guardian:, user:)
+    StaffActionLogger.new(guardian.user).log_username_change(
+      user,
+      user.username_before_last_save,
+      user.username,
+    )
+  end
+end
+```
+And here is an example on how you could call such a service:
+```ruby
+User::UpdateUsername.call(params: {id: 1, username: "username"}, guardian: Discourse.system_user.guardian)
+```
+
+Without knowing how services work, you can probably guess what’s happening here. Let’s dive in.
+
+## Steps
+
+### What’s a step?
+
+This is the basic unit of a service. There is a generic one (`step`) and specialized ones (`params`, `model`, etc.), and they’re all steps.
+
+Steps are defined *in the order they will be called*. Each step will call a corresponding method and, depending on its return value, will continue or halt the execution of the service. Most steps rely on returning a value, and *not raising an exception* (otherwise, you’ll break the execution flow).
+
+The immediate benefit is that error handling is done for you, and you don’t have to implement any specific logic in the service itself. We’ll see later how to handle errors.
+
+> ⚠️ As said above, a step shouldn’t raise an exception, as this will not be handled automatically for you by the service. If a service raises an exception, it should be treated as a bug. If you need to call a piece of code that can raise, then you should provide a `rescue` block for your step to handle the possible exception.
+
+Let’s see what steps are available and how to use them.
+
+### `step`
+
+This is the generic step, you provide a name, and it will run the defined method of the same name. The return value of this step doesn’t impact the execution flow, to mark the service as failed in a generic step, you have to call `#fail!` explicitly.
```

Comment:
Tiny readability suggestion
```suggestion
This is the generic step, you provide a name, and it will run the defined method of the same name. The return value of this step doesn’t impact the execution flow. To mark the service as failed in a generic step, you have to call `#fail!` explicitly.
```